### PR TITLE
LibGfx: Speed up PNGWriter by avoiding some memory copies and allocations

### PR DESCRIPTION
**Pull request by abyesilyurt**
Results on my PC, for 1920x1080 resolution:
- Before: 635 ms
- After: 306 ms

### CODE REVIEW

File: Userland/Libraries/LibGfx/PNGWriter.cpp [hunk 1]
```diff
@@ -13,18 +14,32 @@
 namespace Gfx {
 
 class PNGChunk {
+    using data_length_type = u32;
+
 public:
     explicit PNGChunk(String);
-    Vector<u8> const& data() const { return m_data; };
+    auto const& data() const { return m_data; };
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/PNGWriter.cpp:21]: why not say it's a ByteBuffer

File: Userland/Libraries/LibGfx/PNGWriter.cpp [hunk 2]
```diff
@@ -13,18 +14,32 @@
...
+    void reserve(size_t bytes) { m_data.ensure_capacity(bytes); }
+
+    template<typename T>
+    void add_as_big_endian(T);
+
+    template<typename T>
+    void add_as_little_endian(T);
+
+    template<typename T>
+    void add(T);
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/PNGWriter.cpp:32]: * maybe set some Restrictions on those types
* if you have 'safe' endian aware setters, exposing the raw setter might be redundant and could lead to unsafe code in the future, can you make it private?

File: Userland/Libraries/LibGfx/PNGWriter.cpp [hunk 3]
```diff
@@ -123,23 +146,12 @@ void NonCompressibleBlock::update_adler(u8 data)
     m_adler_s2 = (m_adler_s2 + m_adler_s1) % 65521;
 }
 
-void PNGWriter::add_chunk(PNGChunk const& png_chunk)
+void PNGWriter::add_chunk(PNGChunk& png_chunk)
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/PNGWriter.cpp:149]: You are cahngeing the data the caller might use later, is that inentional?

Comment (abyesilyurt) [reply Userland/Libraries/LibGfx/PNGWriter.cpp:149]: Yes, PNGWriter modifies the content of the png_chunk in this function. I am not sure if there is a cleaner way to do this.

File: Userland/Libraries/LibGfx/PNGWriter.cpp (hunk 2 above)
Comment (abyesilyurt) [reply Userland/Libraries/LibGfx/PNGWriter.cpp]: Added `requires(IsUnsigned<T>)` for add and made it private. Does that work?

File: Userland/Libraries/LibGfx/PNGWriter.cpp (hunk 1 above)
Comment (abyesilyurt) [reply Userland/Libraries/LibGfx/PNGWriter.cpp:22]: Should be good now.

**Pull request merged**